src/common/linear_assignment_cache.h

```diff
@@ -0,0 +1,148 @@
+/*
+
+
+   Copyright 2023 Leil Storage OÜ
+
+   This file is part of SaunaFS.
+
+   SaunaFS is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, version 3.
+
+   SaunaFS is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with SaunaFS. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "common/platform.h"
+
+#include <algorithm>
+#include <array>
+
+#include "common/goal.h"
+
+class LinearAssignmentCache {
+public:
+	static constexpr LinearAssignmentCache *kNotProvidedCache = nullptr;
+
+	class OptimizerInputOutput {
+	public:
+		OptimizerInputOutput()
+		    : target_slice_(Goal::Slice::Type(1)),
+		      source_slice_(Goal::Slice::Type(1)),
+		      assignment_() {}
+
+		static bool partsMatch(const detail::Slice::PartProxy &part1,
+		                       const detail::Slice::PartProxy &part2) {
+			return part1.size() == part2.size() &&
+			       std::equal(part1.begin(), part1.end(), part2.begin());
+		}
+
+		template <std::size_t N>
+		bool checkMatchAndGetResult(Goal::Slice &target_slice,
+		                            Goal::Slice &source_slice,
+		                            std::array<int, N> &assignment) {
+			if (target_slice.getType() != target_slice_.getType() ||
+			    source_slice.getType() != source_slice_.getType()) {
+				return false;
+			}
+			assert(target_slice.size() == target_slice_.size());
```

Comment:
Question: Isn't the same case in the validation of sizes that the validation of types? 
Does it make sense to return false is sizes don't match?

Reply:
The sizes should be same because the types are the same, that is why the assert is here.